rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +399,240 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+        let path_4 = std::path::Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip")
+            .join("Parameters")
+            .join("Interfaces");
+        let path_6 = std::path::Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip6")
+            .join("Parameters")
+            .join("Interfaces");
+        let flags = winreg::enums::KEY_NOTIFY;
+        Ok((
+            hklm.open_subkey_with_flags(path_4, flags)?,
+            hklm.open_subkey_with_flags(path_6, flags)?,
+        ))
+    }
+
+    pub(crate) fn run_debug() -> Result<()> {
+        tracing_subscriber::fmt::init();
+        let rt = tokio::runtime::Runtime::new()?;
+
+        let mut listener = CombinedListener::new()?;
+
+        rt.block_on(async move {
+            loop {
+                tokio::select! {
+                    _r = tokio::signal::ctrl_c() => break,
+                    r = listener.notified() => r?,
+                }
+
+                let resolvers = crate::client::resolvers::get()?;
+                tracing::info!(?resolvers);
+            }
+
+            listener.close()?;
+            Ok::<_, anyhow::Error>(())
+        })?;
+
+        Ok(())
+    }
+
+    pub(crate) struct CombinedListener {
+        listener_4: Listener,
+        listener_6: Listener,
+    }
+
+    impl CombinedListener {
+        pub(crate) fn new() -> Result<Self> {
+            let (key_ipv4, key_ipv6) = open_network_registry_keys()?;
+            let listener_4 = Listener::new(key_ipv4)?;
+            let listener_6 = Listener::new(key_ipv6)?;
+
+            Ok(Self {
+                listener_4,
+                listener_6,
+            })
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            self.listener_4.close()?;
+            self.listener_6.close()?;
+            tracing::debug!("Unregistered registry listener");
+            Ok(())
+        }
+
+        pub(crate) async fn notified(&mut self) -> Result<()> {
+            tokio::select! {
+                r = self.listener_4.notified() => r?,
+                r = self.listener_6.notified() => r?,
+            }
+            Ok(())
+        }
+    }
+
+    /// Listens to one registry key for changes. Callers should await `notified`.
+    struct Listener {
+        key: winreg::RegKey,
+        // Rust never uses `tx` again, but the C callback does. So it must live as long
+        // as Listener, and then be dropped after the C callback is cancelled and any
+        // ongoing callback finishes running.
+        //
+        // We `pin` it here to avoid this sequence:
+        // - `Listener::new` called, pointer to `tx` passed to `RegisterWaitForSingleObject`
+        // - `Listener` and `tx` move
+        // - The callback fires, using the now-invalid previous location of `tx`
+        _tx: std::pin::Pin<Box<mpsc::Sender<()>>>,
```

Comment:
I don't think this needs to be `Pin`, If `Listener` moves, the location pointed to by the `box` wouldn't change since it's heap-allocate, that's why `Box` is `unpin` and why you can still an `&mut T` even when pinned.

Reply:
If this wasn't boxed, `Listener` would need to be `!unpin` but I think that'd be harder to implement without much gain